lib/commands/device/deactivate.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright 2016-2020 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import type { IArg } from '@oclif/parser/lib/args';
+import Command from '../../command';
+import * as cf from '../../utils/common-flags';
+import { getBalenaSdk, stripIndent } from '../../utils/lazy';
+
+interface FlagsDef {
+	yes: boolean;
+	help: void;
+}
+
+interface ArgsDef {
+	uuid: string;
+}
+
+export default class DeviceDeactivateCmd extends Command {
+	public static description = stripIndent`
+		Deactivate a device.
+
+		Deactivate a device.
+
+		Note this command asks for confirmation interactively.
+		You can avoid this by passing the \`--yes\` option.
+		`;
+	public static examples = [
+		'$ balena device deactivate 7cf02a6',
+		'$ balena device deactivate 7cf02a6 --yes',
+	];
+
+	public static args: Array<IArg<any>> = [
+		{
+			name: 'uuid',
+			description: 'the UUID of the device to be deactivated',
+			required: true,
+		},
+	];
+
+	public static usage = 'device deactivate <uuid>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		yes: cf.yes,
+		help: cf.help,
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { args: params, flags: options } = this.parse<FlagsDef, ArgsDef>(
+			DeviceDeactivateCmd,
+		);
+
+		const balena = getBalenaSdk();
+		const patterns = await import('../../utils/patterns');
+
+		const uuid = params.uuid;
+		const deactivationWarning = `
+
+Warning! Deactivating a device will charge a fee equivalent to the
+normal monthly cost for the device (e.g. $1 for an essentials device);
+the device will not be charged again until it comes online.
+
+Are you sure you want to deactivate device ${uuid} ?`;
+
+		// Confirm
+		await patterns.confirm(options.yes, deactivationWarning);
+
+		// Deactivate
+		try {
+			await balena.pine.patch({
+				resource: 'device',
+				body: {
+					is_active: false,
+				},
+				options: {
+					$filter: {
+						uuid: {
+							$startswith: uuid,
```

Comment:
How about doing a `balena.models.device.get(uuid, { $select: 'id' })` right before, so that
* the user gets an error if the device doesn't exist or is ambiguous
* we can use the retrieved `id` directly in the `patch`, instead of a `$startswith` `$filter`

Reply:
Also, @thgreasi is it ok that I am using a pine query here or we should move the `deactivation` to the SDK?

Reply:
My concern is just that $filter $startswith on a PATCH doesn't surface ambiguity issues. I mean, such a patch could apply on more than one devices unintentionally, if it just happens that they share a common uuid prefix and the user passes just that in the CLI. So avoiding $startswith in this case, makes sure that the patch will only apply on a single device, as the command suggests. The cost of the extra GET shouldn't not be a concern, since we also need to surface 404s, which otherwise we would miss.

@mbalamat adding `device.deactivate` to the SDK sounds like a good idea.

Reply:
IMO proper approach would be to add `models.device.disable()` (or something functionally equivalent) to the SDK.

Reply:
> So avoiding $startswith in this case, makes sure that the patch will only apply on a single device, as the command suggests.

Oh yeah. In fact, it would be nice to fetch the full UUID and print the full UUID on the confirmation prompt, e.g.:

```
$ balena device deactivate 59c195a
...
Are you sure you want to deactivate device 59c195afe15206862efc380873e4bdc8 ? (y/N)
```
If the query for the short UUID returns more than one full UUID (ambiguity), the command should aborted, which also covers the `--yes` case.

Reply:
@pdcastro @thgreasi thanks for the suggestions, this https://github.com/balena-io/balena-cli/pull/2112/commits/795d48b4ed921a20f0444c48ff9b8924c781f97d#diff-38726520c30296d8fa2b5e20c4e181ade050b5e9eec92d853b81ddec073fb171R88 will throw a BalenaAmbiguousDevice error if there is any ambiguity, which I am catching. So, the code, as it is right now works only with one device id, and deactivates it, no ambiguity issues.